### PR TITLE
IGNITE-13112 The current security context should be obtained using the IgniteSecurity interface only

### PR DESCRIPTION
**Pull request by dgarus**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestCertificateSecurityProcessor.java [hunk 1]
```diff
@@ -70,14 +73,18 @@ public TestCertificateSecurityProcessor(GridKernalContext ctx, Collection<TestSe
...
+        SecurityContext res = new TestSecurityContext(
             new TestSecuritySubject()
                 .setType(REMOTE_NODE)
                 .setId(node.id())
                 .setAddr(new InetSocketAddress(F.first(node.addresses()), 0))
                 .setLogin("")
                 .setPerms(ALLOW_ALL)
         );
+
+        SECURITY_CONTEXTS.put(res.subject().id(), res);
```
Comment (nizhikov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestCertificateSecurityProcessor.java:85]: Why `SECURITY_CONTEXTS` made static?
AFAICS it used in a non-static context only.

**Comment by nizhikov** [comment]
We need to check and deprecate existing fields that conatins `securitySubjId`, because now it should be transferred in `SecurityAwareCustomMessageWrapper#secSubjId` field.

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java [hunk 2]
```diff
@@ -836,7 +856,8 @@ else if (type == EVT_CLIENT_NODE_RECONNECTED) {
                                         discoCache0,
                                         notification.getTopSnapshot(),
                                         null,
-                                        notification.getSpanContainer()
+                                        notification.getSpanContainer(),
+                                        null
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:860]: Why null here?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java [hunk 3]
```diff
@@ -2233,6 +2259,7 @@ public void metricsUpdateEvent(DiscoCache discoCache, ClusterNode node) {
                 discoCache,
                 discoCache.nodeMap.values(),
                 null,
+                null,
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:2262]: Why null here?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java [hunk 4]
```diff
@@ -2212,6 +2237,7 @@ public void clientCacheStartEvent(UUID reqId,
                     null,
                     Collections.<ClusterNode>emptyList(),
                     new ClientCacheChangeDummyDiscoveryMessage(reqId, startReqs, cachesToClose),
+                    null,
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:2240]: Why null here?

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java [hunk 5]
```diff
@@ -3082,7 +3115,7 @@ else if (log.isDebugEnabled())
                             evt.discoCache = discoCache;
                         }
 
-                        ctx.event().record(customEvt, evt.discoCache);
+                        withContextIfNeed(evt.secSubjId, ctx.security(), () -> ctx.event().record(customEvt, evt.discoCache));
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:3118]: We should change the security context for the whole `DiscoveryWorker`

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java (hunk 2 above)
Comment (dgarus) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:860]: changed to getting of current subject id

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java (hunk 3 above)
Comment (dgarus) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:2262]: changed to getton of current subject id

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java (hunk 4 above)
Comment (dgarus) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:2240]: changed to getton of current subject id

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java [hunk 6]
```diff
@@ -2623,6 +2650,7 @@ public void scheduleSegmentCheck() {
                                     locNodeOnlyTop),
                                 locNodeOnlyTop,
                                 null,
+                                null,
```
[the remark this answers is not in the document]
Comment (dgarus) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:2653]: changed to getting of current subject id

File: modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java (hunk 5 above)
Comment (dgarus) [reply modules/core/src/main/java/org/apache/ignite/internal/managers/discovery/GridDiscoveryManager.java:3118]: ok, done

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheStatisticsModeChangeTask.java [hunk 7]
```diff
@@ -40,6 +43,11 @@ public CacheStatisticsModeChangeTask(CacheStatisticsModeChangeMessage msg) {
         return false;
     }
 
+    /** {@inheritDoc} */
+    @Override public UUID securitySubjectId() {
+        return null;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/CacheStatisticsModeChangeTask.java:48]: This should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/ClientCacheChangeDummyDiscoveryMessage.java [hunk 8]
```diff
@@ -70,6 +73,11 @@ public ClientCacheChangeDummyDiscoveryMessage(
         return true;
     }
 
+    /** {@inheritDoc} */
+    @Override public UUID securitySubjectId() {
+        return null;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/ClientCacheChangeDummyDiscoveryMessage.java:78]: This should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/TxTimeoutOnPartitionMapExchangeChangeTask.java [hunk 9]
```diff
@@ -50,6 +53,11 @@ public TxTimeoutOnPartitionMapExchangeChangeMessage message() {
         return false;
     }
 
+    /** {@inheritDoc} */
+    @Override public UUID securitySubjectId() {
+        return null;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/TxTimeoutOnPartitionMapExchangeChangeTask.java:58]: This should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/WalStateNodeLeaveExchangeTask.java [hunk 10]
```diff
@@ -50,6 +53,11 @@ public ClusterNode node() {
         return true;
     }
 
+    /** {@inheritDoc} */
+    @Override public UUID securitySubjectId() {
+        return null;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/WalStateNodeLeaveExchangeTask.java:58]: This should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/FinishPreloadingTask.java [hunk 11]
```diff
@@ -49,6 +52,11 @@ public FinishPreloadingTask(AffinityTopologyVersion topVer, int grpId) {
         return true;
     }
 
+    /** {@inheritDoc} */
+    @Override public UUID securitySubjectId() {
+        return null;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/FinishPreloadingTask.java:57]: This should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/ForceRebalanceExchangeTask.java [hunk 12]
```diff
@@ -47,6 +50,11 @@ public ForceRebalanceExchangeTask(GridDhtPartitionExchangeId exchId, GridCompoun
         return true;
     }
 
+    /** {@inheritDoc} */
+    @Override public UUID securitySubjectId() {
+        return null;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/ForceRebalanceExchangeTask.java:55]: This should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/RebalanceReassignExchangeTask.java [hunk 13]
```diff
@@ -46,6 +50,11 @@ public RebalanceReassignExchangeTask(GridDhtPartitionExchangeId exchId, GridDhtP
         return true;
     }
 
+    /** {@inheritDoc} */
+    @Override public UUID securitySubjectId() {
+        return null;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/RebalanceReassignExchangeTask.java:55]: This should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/query/schema/SchemaExchangeWorkerTask.java [hunk 14]
```diff
@@ -44,6 +47,11 @@ public SchemaExchangeWorkerTask(SchemaAbstractDiscoveryMessage msg) {
         return false;
     }
 
+    /** {@inheritDoc} */
+    @Override public UUID securitySubjectId() {
+        return null;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/query/schema/SchemaExchangeWorkerTask.java:52]: This should be removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/query/schema/SchemaNodeLeaveExchangeWorkerTask.java [hunk 15]
```diff
@@ -44,6 +47,11 @@ public SchemaNodeLeaveExchangeWorkerTask(ClusterNode node) {
         return true;
     }
 
+    /** {@inheritDoc} */
+    @Override public UUID securitySubjectId() {
+        return null;
```
Comment (nizhikov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/query/schema/SchemaNodeLeaveExchangeWorkerTask.java:52]: This should be removed.

**Comment by bmv126** [comment]
@dgarus 
What is the plan for this PR. Will this be merged into 2.10 master branch ?

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/events/CacheEventsTest.java [hunk 16]
```diff
@@ -0,0 +1,424 @@
...
+        return evtType;
+    }
+
+    /** {@inheritDoc} */
+    @Override protected TestOperation testOperation() {
+        return op;
+    }
+
+    /** */
+    @Override protected GridTestUtils.ConsumerX<String> operation() {
```
Comment (ingvard) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/events/CacheEventsTest.java:340]: java docs.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/events/TaskAndJobEventsTest.java [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+        }
+
+        assertTrue("Remote filter. Events that are not happen: " + rmtSet.stream().map(U::gridEventName),
+            rmtSet.isEmpty());
+        assertTrue("Local listener. Events that are not happen: " + locSet.stream().map(U::gridEventName),
+            locSet.isEmpty());
+    }
+
+    /** */
+    private GridTestUtils.RunnableX operation() {
```
Comment (ingvard) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/events/TaskAndJobEventsTest.java:230]: java docs

File: modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestCertificateSecurityProcessor.java [hunk 18]
```diff
@@ -54,6 +54,9 @@
     /** Permissions. */
     public static final Map<String, SecurityPermissionSet> PERMS = new ConcurrentHashMap<>();
 
+    /** */
+    private final Map<UUID, SecurityContext> secCtxs = new ConcurrentHashMap<>();
```
Comment (ingvard) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/security/impl/TestCertificateSecurityProcessor.java:58]: java docs

File: modules/core/src/test/java/org/apache/ignite/testframework/GridTestUtils.java [hunk 19]
```diff
@@ -2502,6 +2503,29 @@ public static long sleep_and_can_fail() {
...
+     * Consumer that can throw exceptions.
+     */
+    @FunctionalInterface
+    public static interface ConsumerX<T> extends Consumer<T> {
+        /**
+         * Consumer body.
+         *
+         * @throws Exception If failed.
+         */
+        void acceptX(T t) throws Exception;
```
Comment (ingvard) [inline modules/core/src/test/java/org/apache/ignite/testframework/GridTestUtils.java:2516]: public void

**Review by ingvard — CHANGES_REQUESTED** [review]

File: modules/clients/src/test/java/org/apache/ignite/internal/TaskEventSubjectIdSelfTest.java [hunk 20]
```diff
@@ -357,7 +357,7 @@ public void testClient() throws Exception {
         assert evt != null;
 
         assertEquals(EVT_TASK_FINISHED, evt.type());
-        assertEquals(client.id(), evt.subjectId());
+        assertNull(evt.subjectId());
```
[the remark this answers is not in the document]
Comment (dgarus) [reply modules/clients/src/test/java/org/apache/ignite/internal/TaskEventSubjectIdSelfTest.java:360]: This check shows that subjectId has to be null when security is disabled.

File: modules/core/src/main/java/org/apache/ignite/internal/GridKernalContextImpl.java [hunk 21]
```diff
@@ -1343,4 +1349,43 @@ public void recoveryMode(boolean recoveryMode) {
...
+     * Wraps executors to security aware implementations.
+     */
+    private void wrapExecutors() {
+        utilityCachePool = wrapToSecurityAware(utilityCachePool);
+        execSvc = wrapToSecurityAware(execSvc);
+        svcExecSvc = wrapToSecurityAware(svcExecSvc);
+        sysExecSvc = wrapToSecurityAware(sysExecSvc);
+        p2pExecSvc = wrapToSecurityAware(p2pExecSvc);
+        mgmtExecSvc = wrapToSecurityAware(mgmtExecSvc);
+        restExecSvc = wrapToSecurityAware(restExecSvc);
```
[the remark this answers is not in the document]
Comment (dgarus) [reply modules/core/src/main/java/org/apache/ignite/internal/GridKernalContextImpl.java:1351]: No, we should not.
>This code has a strong relation with the initialization order of the executors
I don't sure what you mean. 
GridKernalContextImpl gets all executors through the constructor so they should be initialized for this moment.

**Pull request closed without merging**